### PR TITLE
Always return unknown hint for first sample in non-gauge histogram chunk

### PR DESCRIPTION
**Pull request by fionaliao**
When reading a chunk, always return UnknownCounterReset for the first sample. This is because we can't trust a chunk's counter reset header since we can't always be sure it's really consecutive to the previous chunk being read (e.g. due to backfill or out of order). This is the initial fix for the counter reset detection issue, improvements are being discussed in https://github.com/prometheus/prometheus/issues/15346.

Slack discussion: https://cloud-native.slack.com/archives/C02KR205UMU/p1730898043703779

### CODE REVIEW

File: tsdb/chunkenc/float_histogram.go [hunk 1]
```diff
@@ -752,6 +752,7 @@ func (a *FloatHistogramAppender) AppendFloatHistogram(prev *FloatHistogramAppend
 				a.setCounterResetHeader(NotCounterReset)
 			}
 		}
+
```
Comment (krajorama) [inline tsdb/chunkenc/float_histogram.go:755]:
```suggestion
```

File: tsdb/chunkenc/histogram_meta.go [hunk 2]
```diff
@@ -558,11 +558,12 @@ func counterResetHint(crh CounterResetHeader, numRead uint16) histogram.CounterR
 		// In a counter histogram chunk, there will not be any counter
 		// resets after the first histogram.
 		return histogram.NotCounterReset
-	case crh == CounterReset:
-		// If the chunk was started because of a counter reset, we can
-		// safely return that hint. This histogram always has to be
-		// treated as a counter reset.
-		return histogram.CounterReset
+		// TODO: explain
+	//case crh == CounterReset:
```
Comment (krajorama) [inline tsdb/chunkenc/histogram_meta.go:562]:
```suggestion
	// case crh == CounterReset:
```

**Comment by krajorama** [comment]
Thinking through: on query, we either read the head (in-order/out-of-order) or a block.

If we read the head via chunks (ChunkQuerier):
-  non overlapping chunks are returned as is and when we read them, the hint is gone, so we cannot accidentally invalidate a CounterReset. The rest of the samples are ok to be NotCounterReset as chunks don't overlap.
- overlapping chunks are recoded into new chunks -> wrong NotCounterReset cleared and replaced with not counter reset

If we read head as samples (Querier): I think we end up using ChainSampleIteratorFromIterables which clears the reset hint when switching iterables.

Need to see Block querying and compaction.

**Comment by krajorama** [comment]
One problem could be [remote read](https://github.com/prometheus/prometheus/blob/3c5551df68442dd07668987e0685d12d9c3138dd/prompb/remote.proto#L43) returning chunks. How to tell the client that counter resets are not to be trusted? 
Documentation? After copy to send buffer, clear the header somehow? @fionaliao @beorn7

**Comment by fionaliao** [comment]
Another problem of this implementation of clearing counter resets when reading the first sample of a chunk is that we could be incorrectly setting counter resets when merging blocks (e.g. when backfilling or merging out of order blocks with in-order blocks). This isn't a problem now since we will be ignoring the headers, but later when we fix the counter reset detection properly we won't be able to trust the counter reset headers of older blocks.

File: tsdb/chunkenc/histogram_meta.go [hunk 3]
```diff
@@ -558,11 +558,12 @@ func counterResetHint(crh CounterResetHeader, numRead uint16) histogram.CounterR
...
+	// If the chunk was started because of a counter reset, we can
+	// safely return that hint. This histogram always has to be
+	// treated as a counter reset.
+	//return histogram.CounterReset
 	default:
```
Comment (beorn7) [inline tsdb/chunkenc/histogram_meta.go:567]: The comment below needs a rework based on our current understanding. I also think my TODO below (line 575) is pretty much obsolete.

**Review by beorn7 — COMMENTED** [review]
I like this as the most solid and at the same time easiest solution. Until further notice, let's just always return `histogram.UnknownCounterReset` on every first histogram from a counter histogram chunk.

**Comment by beorn7** [comment]
> One problem could be [remote read](https://github.com/prometheus/prometheus/blob/3c5551df68442dd07668987e0685d12d9c3138dd/prompb/remote.proto#L43) returning chunks. How to tell the client that counter resets are not to be trusted? Documentation? After copy to send buffer, clear the header somehow? @fionaliao @beorn7

I mean, even for chunks coming from remote read, we consider the CRH to be "unknown" for all CRH values that are not "gauge". That should just work, shouldn't it?

**Comment by beorn7** [comment]
> Another problem of this implementation of clearing counter resets when reading the first sample of a chunk is that we could be incorrectly setting counter resets when merging blocks (e.g. when backfilling or merging out of order blocks with in-order blocks). This isn't a problem now since we will be ignoring the headers, but later when we fix the counter reset detection properly we won't be able to trust the counter reset headers of older blocks.

I think we need some specific mechanism that positively affirms if two chunks that are now consecutive had been consecutive during creation of the 2nd chunk. Only then will we trust the CRH. In that way, we should never get it wrong.

**Comment by beorn7** [comment]
> Need to think whether this covers all the read cases

Yes indeed. I would assume it all goes through iterators, so we should be able to find out about it.

**Comment by krajorama** [comment]
> > One problem could be [remote read](https://github.com/prometheus/prometheus/blob/3c5551df68442dd07668987e0685d12d9c3138dd/prompb/remote.proto#L43) returning chunks. How to tell the client that counter resets are not to be trusted? Documentation? After copy to send buffer, clear the header somehow? @fionaliao @beorn7
> 
> I mean, even for chunks coming from remote read, we consider the CRH to be "unknown" for all CRH values that are not "gauge". That should just work, shouldn't it?

I meant the other way around when we are the remote read server, sending out chunks.

To me it seems optional whether we merge this PR or not. I think it's actually more important to try to prevent writing out CRH that's false, to have a chance of progressing later towards a better solution and not have too much corrupt data out there. See CNCF slack discussion and plan proposal.

**Comment by beorn7** [comment]
> I meant the other way around when we are the remote read server, sending out chunks.

We could send them out with CRH unknown, or we could do something else, but the spec is equally valid for the remote read receiver, which also must ignore CRH for counter histogram chunks.

> I think it's actually more important to try to prevent writing out CRH that's false, to have a chance of progressing later towards a better solution and not have too much corrupt data out there. 

Right now, even the "correct" CRHs cannot be trusted, so it's - in a way - fairly irrelevant what we write right now. Only the distinction gauge vs. counter is relevant right now.

File: tsdb/chunkenc/histogram_meta.go (hunk 3 above)
Comment (fionaliao) [reply tsdb/chunkenc/histogram_meta.go:561]: Reworked the comment in https://github.com/prometheus/prometheus/pull/15343/commits/8311877dfc99e1ca9ab56a9a13dad8db05652041

**Comment by fionaliao** [comment]
I added the tests from https://github.com/prometheus/prometheus/pull/15221 and https://github.com/prometheus/prometheus/pull/15253 to this PR

File: tsdb/chunkenc/histogram_meta.go (hunk 3 above)
Comment (fionaliao) [reply tsdb/chunkenc/histogram_meta.go:561]: Also fixed a bit in https://github.com/prometheus/prometheus/pull/15343/commits/01717673255f7ad573200d3f3cb06f4807eab7b2

**Review by krajorama — APPROVED** [review]
LGTM

**Review by beorn7 — APPROVED** [review]
Thank you very much (for this PR and all the productive and insightful discussion).

**Pull request merged**